File: 코딩테스트/사각형.py
from math import factorial


D = [[19,17]]
R = 37
C = 51

'''
2<= N <= 250
y+x C x
'''

sum = [[0 for _ in range(101)] for _ in range(101)]

def Comb(n : int,r : int)->int:
    global sum
    
    if sum[n][r]:
        return sum[n][r]
    
    if n == r or r == 0:
        return 1
    else:
        sum[n][r] = Comb(n-1,r-1) + Comb(n-1,r)
        return sum[n][r]

# log(N)

def reset():
    global sum

    sum = [[0 for _ in range(101)] for _ in range(101)]

def Comb2(n : int, r : int)->int:
    return factorial(n+r)%10000019 / ((factorial(n-r) * factorial(r)))

if __name__ == "__main__":
    answer = 0
    for [y,x] in D:
        # 왼쪽 위에서 아래로 갈 때
        
        reset()
        a = Comb(y+x-1, y)
        reset()
        b = Comb(R+C-(x+y-1), R-y-1)

        answer += (a * b) % 10000019

        # 오른쪽 아래에서 위로 갈 때
        reset()
        a = Comb(y+x-1, x)
        reset()
        b = Comb(R+C-(x+y-1), C-x-1)
        answer += (a*b) % 10000019

    print(answer)

File: 코딩테스트/사각형.cpp
#include <iostream>
using namespace std;

int D[][2] = {{17,19}};     //대각선의 위치
int R=51, C=37;                       //세로, 가로 크기

long long sum[100][100] = {0};

long long Comb(int n, int r)
{
    if(sum[n][r] != 0) return sum[n][r];

    if(n == r || r == 0)
        return 1;
    else
        sum[n][r] = Comb(n-1,r-1) % 1000019 + Comb(n-1,r)%1000019;
        return sum[n][r];
}

long long Solution()
{
    long long answer = 0;

    for(const auto[y,x] : D)
    {
        answer += (Comb(y+x, x) * Comb(R+C-y, R-y))%1000019;
    }

    return answer;
}

int main()
{
    cout<<Solution()<<endl;
}

/*
    (R+C)Comb(C) = (R+C)! / (R! * C!)
    1 * 3!/2! = 6
*/